File: tools-manage/invocationlab-admin/src/main/java/org/javamaster/invocationlab/admin/service/impl/RedisServiceImpl.java
package org.javamaster.invocationlab.admin.service.impl;

import com.google.common.collect.Lists;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.RandomStringUtils;
import org.javamaster.invocationlab.admin.config.BizException;
import org.javamaster.invocationlab.admin.model.erd.TokenVo;
import org.javamaster.invocationlab.admin.model.redis.CommonRedisVo;
import org.javamaster.invocationlab.admin.model.redis.ConnectionVo;
import org.javamaster.invocationlab.admin.model.redis.Tree;
import org.javamaster.invocationlab.admin.model.redis.ValueVo;
import org.javamaster.invocationlab.admin.redis.RedisDataTypeService;
import org.javamaster.invocationlab.admin.redis.RedisHelper;
import org.javamaster.invocationlab.admin.redis.TripleFunction;
import org.javamaster.invocationlab.admin.service.Pair;
import org.javamaster.invocationlab.admin.service.RedisService;
import org.javamaster.invocationlab.admin.util.RedisUtils;
import org.javamaster.invocationlab.admin.util.SpringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.redis.connection.DataType;
import org.springframework.data.redis.connection.RedisConnection;
import org.springframework.data.redis.connection.jedis.JedisClusterConnection;
import org.springframework.data.redis.connection.jedis.JedisConnectionFactory;
import org.springframework.data.redis.core.RedisCallback;
import org.springframework.data.redis.core.RedisConnectionUtils;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Service;
import org.springframework.web.context.request.RequestContextHolder;
import org.springframework.web.context.request.ServletRequestAttributes;
import redis.clients.jedis.Jedis;
import redis.clients.jedis.JedisCluster;
import redis.clients.jedis.Protocol;

import javax.servlet.http.HttpSession;
import java.nio.charset.StandardCharsets;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import static org.javamaster.invocationlab.admin.util.RedisUtils.getConnectionVo;

/**
 * @author yudong
 */
@Slf4j
@Service
public class RedisServiceImpl implements RedisService {
    public static final String HASH_KEY_DBS = "admin:erd:redis:dbs";
    @Autowired
    private RedisHelper redisHelper;
    @Autowired
    private RedisTemplate<String, Object> redisTemplateJackson;
    @Autowired
    private Map<String, RedisDataTypeService> redisDataTypeServiceMap;

    @Override
    public String pingConnect(ConnectionVo connectionVoReq) {
        if (connectionVoReq.getHost().contains(",")) {
            connectionVoReq.setNodes(connectionVoReq.getHost());
            connectionVoReq.setHost("");
        }
        JedisConnectionFactory factory = RedisUtils.newJedisConnectionFactory(connectionVoReq, Protocol.DEFAULT_DATABASE);
        RedisConnection connection = null;
        try {
            connection = factory.getConnection();
        } finally {
            if (connection != null) {
                connection.close();
            }
            factory.destroy();
        }
        return connectionVoReq.getHost() + "连接成功";
    }

    @Override
    public String saveConnect(ConnectionVo connectionVoReq) {
        if (connectionVoReq.getHost().contains(",")) {
            connectionVoReq.setNodes(connectionVoReq.getHost());
            connectionVoReq.setHost("");
        }
        String connectId = RandomStringUtils.randomAlphabetic(12);
        connectionVoReq.setConnectId(connectId);
        connectionVoReq.setCreateTime(System.currentTimeMillis());
        redisTemplateJackson.opsForHash().put(HASH_KEY_DBS, connectId, connectionVoReq);
        return "保存成功";
    }

    @Override
    public List<ConnectionVo> listConnects() {
        List<Object> values = redisTemplateJackson.opsForHash().values(HASH_KEY_DBS);
        return values.stream()
                .map(obj -> {
                    ConnectionVo connectionVo = (ConnectionVo) obj;
                    connectionVo.setPassword("");
                    connectionVo.setUser("");
                    return connectionVo;
                })
                .sorted(Comparator.comparing(ConnectionVo::getCreateTime))
                .collect(Collectors.toList());
    }

    @Override
    public List<Tree> listDb(String connectId) {
        RedisTemplate<Object, Object> redisTemplate = RedisUtils.redisTemplateSingleton(connectId, Protocol.DEFAULT_DATABASE);
        return redisTemplate.execute((RedisCallback<List<Tree>>) connection -> {
            Object redisConnection = connection.getNativeConnection();
            Pair<Integer, Map<Integer, Long>> pair;
            if (redisConnection instanceof Jedis) {
                Jedis jedis = (Jedis) redisConnection;
                pair = redisHelper.getSingleDbs(jedis);
            } else {
                JedisCluster jedisCluster = (JedisCluster) redisConnection;
                pair = redisHelper.getClusterDbs(jedisCluster);
            }
            Integer dbCount = pair.getLeft();
            Map<Integer, Long> map = pair.getRight();
            return IntStream.range(0, dbCount)
                    .mapToObj(index -> {
                        Long keyCount = map.get(index);
                        return Tree.builder()
                                .redisDbIndex(index)
                                .keyCount(keyCount)
                                .isLeaf(false)
                                .label("DB" + index + "(" + keyCount + ")")
                                .build();
                    })
                    .collect(Collectors.toList());
        });
    }

    @Override
    public List<Tree> listKeys(String connectId, Integer redisDbIndex, String pattern) {
        int count = 10000;
        if (pattern.equals("*")) {
            count = 100;
        }
        int finalCount = count;
        RedisTemplate<Object, Object> redisTemplate = RedisUtils.redisTemplateSingleton(connectId, redisDbIndex);
        List<Tree> resList = redisTemplate.execute((RedisCallback<List<Tree>>) connection -> {
            List<Tree> list = Lists.newArrayList();

            if (!pattern.contains("*") && !pattern.contains("?")) {
                byte[] patternBytes = pattern.getBytes(StandardCharsets.UTF_8);
                Boolean exists = connection.exists(patternBytes);
                if (Boolean.TRUE.equals(exists)) {
                    @SuppressWarnings("ConstantConditions")
                    String type = RedisUtils.handleType(connection.type(patternBytes).code());
                    Tree tree = Tree.builder()
                            .label(type + pattern)
                            .labelBase64("")
                            .typeLength(type.length())
                            .isLeaf(true)
                            .build();
                    list.add(tree);
                }
                return list;
            }

            Object nativeConnection = connection.getNativeConnection();
            if (nativeConnection instanceof Jedis) {
                Jedis jedis = (Jedis) nativeConnection;
                list = redisHelper.getSingleDbKeys(jedis, finalCount, pattern);
            } else {
                JedisClusterConnection clusterConnection = (JedisClusterConnection) RedisConnectionUtils
                        .getConnection(Objects.requireNonNull(redisTemplate.getConnectionFactory()));
                ConnectionVo connectionVo = getConnectionVo(connectId);
                list = redisHelper.getClusterDbKeys(clusterConnection, connectionVo, finalCount, pattern);
            }

            return list;
        });

        //noinspection ConstantConditions
        if (resList.size() > 200) {
            resList = resList.subList(0, 200);
        }

        resList.sort(Comparator.comparing(Tree::getLabel));

        return resList;
    }

    @Override
    public ValueVo getValue(CommonRedisVo commonRedisVo) throws Exception {
        return executeCommand(commonRedisVo, (connection, keyPair, redisDataTypeService)
                -> redisDataTypeService.getValue(connection, keyPair));
    }

    @Override
    public ValueVo saveValue(CommonRedisVo commonRedisVo) throws Exception {
        TokenVo tokenVo = getTokenVo();
        log.info("{}-{} save value:{}", tokenVo.getUserId(), tokenVo.getUsername(), commonRedisVo);

        return executeCommand(commonRedisVo, (connection, keyPair, redisDataTypeService) -> {
            Long aLong = redisDataTypeService.saveValue(connection, keyPair, commonRedisVo);
            log.info("保存结果:{}", aLong);
            return commonRedisVo;
        });
    }

    @Override
    public String delField(CommonRedisVo commonRedisVo) throws Exception {
        TokenVo tokenVo = getTokenVo();
        log.info("{}-{} del field:{}", tokenVo.getUserId(), tokenVo.getUsername(), commonRedisVo);

        return executeCommand(commonRedisVo, (connection, keyPair, redisDataTypeService) -> {
            Long affect = redisDataTypeService.delField(connection, keyPair, commonRedisVo);
            return "删除成功:" + affect;
        });
    }

    @Override
    public String addField(CommonRedisVo commonRedisVo) throws Exception {
        TokenVo tokenVo = getTokenVo();
        log.info("{}-{} add field:{}", tokenVo.getUserId(), tokenVo.getUsername(), commonRedisVo);

        return executeCommand(commonRedisVo, (connection, keyPair, redisDataTypeService) -> {
            String affect = redisDataTypeService.addField(connection, keyPair, commonRedisVo);
            return "新增成功:" + affect;
        });
    }

    @Override
    public String setNewTtl(CommonRedisVo commonRedisVo) {
        TokenVo tokenVo = getTokenVo();
        log.info("{}-{} set new ttl:{}", tokenVo.getUserId(), tokenVo.getUsername(), commonRedisVo);

        return executeCommand(commonRedisVo, (connection, keyPair, redisDataTypeService) -> {
            Boolean res = redisDataTypeService.setTtlIfNecessary(connection, keyPair, commonRedisVo.getRedisKeyTtl());
            return "设置结果:" + res;
        });
    }

    @Override
    public ValueVo addKey(CommonRedisVo commonRedisVo) {
        TokenVo tokenVo = getTokenVo();
        log.info("{}-{} add key:{}", tokenVo.getUserId(), tokenVo.getUsername(), commonRedisVo);

        RedisTemplate<Object, Object> redisTemplate = RedisUtils.redisTemplateSingleton(commonRedisVo.getConnectId(),
                commonRedisVo.getRedisDbIndex());

        Pair<byte[], Class<?>> keyPair = redisHelper.convertKeyToBytes(commonRedisVo.getRedisKey(),
                commonRedisVo.getRedisKeyJdkSerialize());
        byte[] keyBytes = keyPair.getLeft();

        return redisTemplate.execute((RedisCallback<ValueVo>) connection -> {
            if (Boolean.TRUE.equals(connection.keyCommands().exists(keyBytes))) {
                throw new BizException("Redis key: " + commonRedisVo.getRedisKey() + " 已存在!!!");
            }

            DataType dataType = DataType.fromCode(commonRedisVo.getRedisKeyType());
            RedisDataTypeService redisDataTypeService = redisDataTypeServiceMap.get(dataType.code());
            ValueVo valueVo = redisDataTypeService.addKey(connection, keyPair, commonRedisVo);

            redisDataTypeService.setTtlIfNecessary(connection, keyPair, commonRedisVo.getRedisKeyTtl());

            return valueVo;
        });
    }

    @Override
    public String delKey(CommonRedisVo commonRedisVo) {
        TokenVo tokenVo = getTokenVo();
        log.info("{}-{} del key:{}", tokenVo.getUserId(), tokenVo.getUsername(), commonRedisVo);

        return executeCommand(commonRedisVo, (connection, keyPair, redisDataTypeService) ->
                "删除结果:" + redisDataTypeService.delKey(connection, keyPair));
    }

    @Override
    public ValueVo renameKey(CommonRedisVo commonRedisVo) {
        TokenVo tokenVo = getTokenVo();
        log.info("{}-{} rename key:{}", tokenVo.getUserId(), tokenVo.getUsername(), commonRedisVo);

        RedisTemplate<Object, Object> redisTemplate = RedisUtils.redisTemplateSingleton(commonRedisVo.getConnectId(),
                commonRedisVo.getRedisDbIndex());

        Pair<byte[], Class<?>> keyPair = redisHelper.convertKeyToBytes(commonRedisVo.getRedisKey(),
                commonRedisVo.getRedisKeyJdkSerialize());
        Pair<byte[], Class<?>> oldKeyPair = redisHelper.convertKeyToBytes(commonRedisVo.getOldRedisKey(),
                commonRedisVo.getRedisKeyBase64());

        return redisTemplate.execute((RedisCallback<ValueVo>) connection -> {
            DataType dataType = connection.keyCommands().type(oldKeyPair.getLeft());
            if (dataType == null || dataType == DataType.NONE) {
                throw new BizException("Redis key: " + commonRedisVo.getOldRedisKey() + " 已不存在!!!");
            }

            RedisDataTypeService redisDataTypeService = redisDataTypeServiceMap.get(dataType.code());
            return redisDataTypeService.renameKey(connection, oldKeyPair, keyPair, commonRedisVo);
        });
    }

    public <U> U executeCommand(CommonRedisVo commonRedisVo,
                                TripleFunction<RedisConnection, Pair<byte[], Class<?>>, RedisDataTypeService, U> function) {
        RedisTemplate<Object, Object> redisTemplate = RedisUtils.redisTemplateSingleton(commonRedisVo.getConnectId(),
                commonRedisVo.getRedisDbIndex());

        RedisHelper redisHelper = SpringUtils.getContext().getBean(RedisHelper.class);
        Pair<byte[], Class<?>> keyPair = redisHelper.convertKeyToBytes(commonRedisVo.getRedisKey(),
                commonRedisVo.getRedisKeyBase64());
        byte[] keyBytes = keyPair.getLeft();

        return redisTemplate.execute((RedisCallback<U>) connection -> {
            DataType dataType = connection.keyCommands().type(keyBytes);
            if (dataType == null || dataType == DataType.NONE) {
                throw new BizException("Redis key: " + commonRedisVo.getRedisKey() + " 已不存在!!!");
            }

            RedisDataTypeService redisDataTypeService = redisDataTypeServiceMap.get(dataType.code());
            return function.apply(connection, keyPair, redisDataTypeService);
        });
    }

    public static TokenVo getTokenVo() {
        ServletRequestAttributes requestAttributes = (ServletRequestAttributes) RequestContextHolder.getRequestAttributes();
        @SuppressWarnings("ConstantConditions")
        HttpSession session = requestAttributes.getRequest().getSession();
        TokenVo tokenVo = (TokenVo) session.getAttribute("tokenVo");
        if (tokenVo == null) {
            tokenVo = new TokenVo();
            tokenVo.setUserId("anonymous");
            tokenVo.setUsername("匿名");
        }
        return tokenVo;
    }
}

File: tools-manage/invocationlab-admin-21/src/main/kotlin/org/javamaster/invocationlab/admin/service/impl/RedisServiceImpl.kt
package org.javamaster.invocationlab.admin.service.impl

import org.javamaster.invocationlab.admin.config.BizException
import org.javamaster.invocationlab.admin.model.erd.TokenVo
import org.javamaster.invocationlab.admin.model.redis.CommonRedisVo
import org.javamaster.invocationlab.admin.model.redis.ConnectionVo
import org.javamaster.invocationlab.admin.model.redis.Tree
import org.javamaster.invocationlab.admin.model.redis.ValueVo
import org.javamaster.invocationlab.admin.redis.RedisHelper
import org.javamaster.invocationlab.admin.redis.RedisStrategy
import org.javamaster.invocationlab.admin.redis.RedisStrategy.Companion.getInstance
import org.javamaster.invocationlab.admin.redis.TripleFunction
import org.javamaster.invocationlab.admin.service.Pair
import org.javamaster.invocationlab.admin.service.RedisService
import org.javamaster.invocationlab.admin.util.RedisUtils.getConnectionVo
import org.javamaster.invocationlab.admin.util.RedisUtils.getJedisConnectionFactory
import org.javamaster.invocationlab.admin.util.RedisUtils.getRedisTemplate
import org.javamaster.invocationlab.admin.util.SpringUtils
import org.apache.commons.lang3.RandomStringUtils
import org.apache.commons.lang3.tuple.Triple
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.redis.connection.DataType
import org.springframework.data.redis.connection.RedisConnection
import org.springframework.data.redis.connection.jedis.JedisClusterConnection
import org.springframework.data.redis.core.RedisCallback
import org.springframework.data.redis.core.RedisConnectionUtils
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.stereotype.Service
import org.springframework.web.context.request.RequestContextHolder
import org.springframework.web.context.request.ServletRequestAttributes
import redis.clients.jedis.Jedis
import redis.clients.jedis.JedisCluster
import redis.clients.jedis.Protocol
import java.nio.charset.StandardCharsets
import java.util.*
import java.util.stream.Collectors
import java.util.stream.IntStream

/**
 * @author yudong
 */

@Service
class RedisServiceImpl : RedisService {
    val log: Logger = LoggerFactory.getLogger(this.javaClass)

    @Autowired
    private lateinit var redisHelper: RedisHelper

    @Autowired
    private val redisTemplateJackson: RedisTemplate<String, Any>? = null

    override fun pingConnect(connectionVoReq: ConnectionVo): String {
        if (connectionVoReq.host!!.contains(",")) {
            connectionVoReq.nodes = connectionVoReq.host
            connectionVoReq.host = ""
        }
        val factory = getJedisConnectionFactory(connectionVoReq, Protocol.DEFAULT_DATABASE)
        factory.connection.use { }
        return connectionVoReq.host + "连接成功"
    }

    override fun saveConnect(connectionVoReq: ConnectionVo): String {
        if (connectionVoReq.host!!.contains(",")) {
            connectionVoReq.nodes = connectionVoReq.host
            connectionVoReq.host = ""
        }
        val connectId = RandomStringUtils.randomAlphabetic(12)
        connectionVoReq.connectId = connectId
        connectionVoReq.createTime = System.currentTimeMillis()
        redisTemplateJackson!!.opsForHash<Any, Any>().put(HASH_KEY_DBS, connectId, connectionVoReq)
        return "保存成功"
    }

    override fun listConnects(): List<ConnectionVo> {
        val values = redisTemplateJackson!!.opsForHash<Any, Any>().values(HASH_KEY_DBS)
        return values.stream()
            .map { obj: Any ->
                val connectionVo = obj as ConnectionVo
                connectionVo.password = ""
                connectionVo.user = ""
                if (connectionVo.createTime == null) {
                    connectionVo.createTime = System.currentTimeMillis()
                }
                connectionVo
            }
            .sorted { a, b ->
                a.createTime!!.compareTo(b.createTime!!)
            }
            .collect(Collectors.toList())
    }

    override fun listDb(connectId: String): List<Tree> {
        val redisTemplate = getRedisTemplate(connectId, Protocol.DEFAULT_DATABASE)
        return redisTemplate.execute { connection: RedisConnection ->
            val redisConnection = connection.nativeConnection
            val pair: Pair<Int, Map<Int, Long>>
            if (redisConnection is Jedis) {
                pair = redisHelper.getSingleDbs(redisConnection)
            } else {
                val jedisCluster = redisConnection as JedisCluster
                pair = redisHelper.getClusterDbs(jedisCluster, redisTemplate)
            }
            val dbCount = pair.left
            val map = pair.right
            IntStream.range(0, dbCount)
                .mapToObj { index: Int ->
                    Tree().apply {
                        redisDbIndex = index
                        keyCount = -1L
                        isLeaf = false
                        label = "DB" + index + "(" + map[index] + ")"
                    }
                }
                .filter { obj: Any? -> Objects.nonNull(obj) }
                .collect(Collectors.toList()) as List<Tree>
        }!!
    }

    override fun listKeys(connectId: String, redisDbIndex: Int, pattern: String): List<Tree> {
        var count = 10000
        if (pattern == "*") {
            count = 100
        }
        val redisTemplate = getRedisTemplate(connectId, redisDbIndex)
        val finalCount = count
        var resList = redisTemplate.execute(
            RedisCallback { connection: RedisConnection ->
                var list: List<Triple<String, String, Class<*>>> = arrayListOf()
                if (!pattern.contains("*") && !pattern.contains("")) {
                    val exists = connection.keyCommands().exists(pattern.toByteArray(StandardCharsets.UTF_8))
                    if (exists == true) {
                        list.addLast(Triple.of(pattern, "", null))
                    }
                    return@RedisCallback list
                }

                val nativeConnection = connection.nativeConnection
                if (nativeConnection is Jedis) {
                    list = redisHelper.getSingleDbKeys(nativeConnection, finalCount, pattern)
                } else {
                    val clusterConnection = RedisConnectionUtils
                        .getConnection(redisTemplate.connectionFactory!!) as JedisClusterConnection
                    val connectionVo = getConnectionVo(connectId)
                    list = redisHelper.getClusterDbKeys(clusterConnection, connectionVo, finalCount, pattern)
                }
                list
            })!!

        if (Objects.requireNonNull(resList).size > 200) {
            resList = resList.subList(0, 200)
        }

        return resList.stream()
            .sorted(
                Comparator.comparing { obj: Triple<String, String, Class<*>> -> obj.left }
            )
            .map { triple: Triple<String, String, Class<*>> ->
                Tree().apply {
                    label = triple.left
                    labelBase64 = triple.middle
                    isLeaf = true
                }
            }
            .collect(Collectors.toList())
    }


    override fun getValue(commonRedisVo: CommonRedisVo): ValueVo {
        return executeCommand(commonRedisVo) { connection: RedisConnection, keyPair: Pair<ByteArray, Class<*>>, redisStrategy: RedisStrategy ->
            redisStrategy.getValue(
                connection, keyPair
            )
        }
    }


    override fun saveValue(commonRedisVo: CommonRedisVo): ValueVo {
        val tokenVo = tokenVo
        log.info("{}-{} save value:{}", tokenVo.userId, tokenVo.username, commonRedisVo)

        return executeCommand(commonRedisVo) { connection: RedisConnection, keyPair: Pair<ByteArray, Class<*>>, redisStrategy: RedisStrategy ->
            redisStrategy.saveValue(
                connection, keyPair, commonRedisVo
            )
            commonRedisVo
        }
    }


    override fun delField(commonRedisVo: CommonRedisVo): String {
        val tokenVo = tokenVo
        log.info("{}-{} del field:{}", tokenVo.userId, tokenVo.username, commonRedisVo)

        return executeCommand(commonRedisVo) { connection: RedisConnection, keyPair: Pair<ByteArray, Class<*>>, redisStrategy: RedisStrategy ->
            val affect = redisStrategy.delField(
                connection, keyPair, commonRedisVo
            )
            "删除成功:$affect"
        }
    }


    override fun addField(commonRedisVo: CommonRedisVo): String {
        log.info("{}-{} add field:{}", tokenVo.userId, tokenVo.username, commonRedisVo)

        return executeCommand(commonRedisVo) { connection, keyPair, redisStrategy ->
            val affect = redisStrategy.addField(connection, keyPair, commonRedisVo)
            "新增成功:$affect"
        }
    }

    override fun setNewTtl(commonRedisVo: CommonRedisVo): String {
        val tokenVo = tokenVo
        log.info("{}-{} set new ttl:{}", tokenVo.userId, tokenVo.username, commonRedisVo)

        return executeCommand(commonRedisVo) { connection: RedisConnection, keyPair: Pair<ByteArray, Class<*>>, redisStrategy: RedisStrategy ->
            val res = redisStrategy.setTtlIfNecessary(
                connection, keyPair, commonRedisVo.redisKeyTtl!!
            )
            "设置结果:$res"
        }
    }

    override fun addKey(commonRedisVo: CommonRedisVo): ValueVo {
        val tokenVo = tokenVo
        log.info("{}-{} add key:{}", tokenVo.userId, tokenVo.username, commonRedisVo)

        val redisTemplate = getRedisTemplate(commonRedisVo.connectId!!, commonRedisVo.redisDbIndex!!)

        val keyPair = redisHelper.convertKeyToBytes(
            commonRedisVo.redisKey!!, commonRedisVo.redisKeyJdkSerialize!!
        )
        val keyBytes = keyPair.left

        return redisTemplate.execute<ValueVo> { connection: RedisConnection ->
            if (true == connection.keyCommands().exists(keyBytes)) {
                throw BizException("Redis key: " + commonRedisVo.redisKey + " 已存在!!!")
            }
            val dataType = DataType.fromCode(commonRedisVo.redisKeyType!!)
            val redisStrategy = getInstance(dataType)
            val valueVo = redisStrategy.addKey(connection, keyPair, commonRedisVo)

            redisStrategy.setTtlIfNecessary(connection, keyPair, commonRedisVo.redisKeyTtl!!)
            valueVo
        }!!
    }

    override fun delKey(commonRedisVo: CommonRedisVo): String {
        val tokenVo = tokenVo
        log.info("{}-{} del key:{}", tokenVo.userId, tokenVo.username, commonRedisVo)

        return executeCommand(commonRedisVo) { connection: RedisConnection, keyPair: Pair<ByteArray, Class<*>>, redisStrategy: RedisStrategy ->
            "删除结果:" + redisStrategy.delKey(
                connection, keyPair
            )
        }
    }

    override fun renameKey(commonRedisVo: CommonRedisVo): ValueVo {
        val tokenVo = tokenVo
        log.info("{}-{} rename key:{}", tokenVo.userId, tokenVo.username, commonRedisVo)

        val redisTemplate = getRedisTemplate(commonRedisVo.connectId!!, commonRedisVo.redisDbIndex!!)

        val keyPair = redisHelper.convertKeyToBytes(
            commonRedisVo.redisKey!!, commonRedisVo.redisKeyJdkSerialize!!
        )
        val oldKeyPair = redisHelper.convertKeyToBytes(
            commonRedisVo.oldRedisKey!!,
            commonRedisVo.redisKeyBase64!!
        )

        return redisTemplate.execute { connection: RedisConnection ->
            val dataType = connection.keyCommands().type(oldKeyPair.left)
            if (dataType == DataType.NONE) {
                throw BizException("Redis key: " + commonRedisVo.oldRedisKey + " 已不存在!!!")
            }

            val redisStrategy = getInstance(dataType!!)
            redisStrategy.renameKey(connection, oldKeyPair, keyPair, commonRedisVo)
        }!!
    }

    companion object {
        const val HASH_KEY_DBS: String = "erd:redis:dbs"
        fun <U> executeCommand(
            commonRedisVo: CommonRedisVo,
            function: TripleFunction<RedisConnection, Pair<ByteArray, Class<*>>, RedisStrategy, U>
        ): U {
            val redisTemplate = getRedisTemplate(commonRedisVo.connectId!!, commonRedisVo.redisDbIndex!!)

            val redisHelper = SpringUtils.context.getBean(RedisHelper::class.java)
            val keyPair: Pair<ByteArray, Class<*>> = redisHelper.convertKeyToBytes(
                commonRedisVo.redisKey!!, commonRedisVo.redisKeyBase64!!
            )
            val keyBytes = keyPair.left

            return redisTemplate.execute { connection: RedisConnection ->
                val dataType = connection.keyCommands().type(
                    keyBytes
                )
                if (dataType == DataType.NONE) {
                    throw BizException("Redis key: " + commonRedisVo.redisKey + " 已不存在!!!")
                }

                val redisStrategy = getInstance(dataType!!)
                function.apply(connection, keyPair, redisStrategy)
            }!!
        }

        @JvmStatic
        val tokenVo: TokenVo
            get() {
                val requestAttributes = RequestContextHolder.getRequestAttributes() as ServletRequestAttributes
                val session = requestAttributes.request.session
                var tokenVo = session.getAttribute("tokenVo") as TokenVo?
                if (tokenVo == null) {
                    tokenVo = TokenVo()
                    tokenVo.userId = "anonymous"
                    tokenVo.username = "匿名"
                }
                return tokenVo
            }
    }
}
